### PR TITLE
Allow e.com to link straight to e.cash and share sessions

### PR DESCRIPTION
**Pull request by tgolen**
### Fixed Issues
Done along with: https://github.com/Expensify/Web-Expensify/pull/31302
Fixes https://github.com/Expensify/Expensify/issues/162554

### Tests / QA
1. Be sure you're signed out of e.cash
2. Sign into e.com with a brand new account on a domain WITHOUT two-factor authentication
3. Complete the inbox task "What will you use Expensify for? Choose as many as you'd like!" and ONLY choose the expensify card
4. Verify that you are automatically logged into e.cash and are seeing the page to create a new workspace

Now, test it out with 2FA

1. Be sure you're signed out of e.cash
2. Sign into e.com with a brand new account on a domain WITH two-factor authentication enabled
3. Complete the inbox task "What will you use Expensify for? Choose as many as you'd like!" and ONLY choose the expensify card
4. Verify that you are redirected to e.cash to ever a 2FA code
5. After the code is entered, verify you are automatically logged into e.cash and are seeing the page to create a new workspace

### Tested On
No other platforms need to be tested because this flow only works on web (the inbox task isn't shown on mobile at all)
- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

#### Web

image

cc @marcaaron @TomatoToaster

### CODE REVIEW

**Comment by tgolen** [comment]
This is all working, but the redirect to the new-workspace page is causing several JS console warnings:

image

They don't appear to be affecting anything, but I haven't been able to figure out what the deal is. My guess is that it's related to running these back-to-back:

```
            Navigation.dismissModal();
            Navigation.navigate(ROUTES.WORKSPACE_NEW);
```

and I think there is a race condition because I think part of the code inside `dismissModal()` is async, but it's hard to figure out when exactly `dismissModal()` is finished doing its thing. There is one existing flow that uses this same dismiss->navigate pattern here: https://github.com/Expensify/Expensify.cash/blob/099457ec41f238be8103d953e4c8b50ea2b030f1/src/libs/actions/Policy.js#L201-L202

I'm not sure what to do about this. We need to have this code deployed by Jul 9, so I am thinking that we merge this as-is, and then open up an issue for improving this dismiss->navigate pattern.

**Review by marcaaron — COMMENTED** [review]
These changes LGTM! I'm not sure whether there is any problem with the dismiss + navigate pattern (it seems kosher to me). But I agree that the react warning is not a blocker. Could be an issue is `withOnyx()` not having any awareness that its children have unmounted - I've observed similar issues before, but hard to say without debugging.

File: src/pages/ValidateLogin2FANewWorkspacePage.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            loading: false,
+        };
+    }
+
+    componentDidMount() {
+        // If the user has an active session already, they need to be redirected straight to the new workspace page
+        if (this.props.session.authToken) {
+            // In order to navigate to a modal, we first have to dismiss the current modal. But there is no current
+            // modal you say? I know, it confuses me too. Without dismissing the current modal, if they user cancels
+            // out of the new workspace modal, then they will be routed back to
```
[the remark this answers is not in the document]
Comment (tgolen) [reply src/pages/ValidateLogin2FANewWorkspacePage.js:59]: Yeah, I'm all in favor of updating this comment to make more sense. If the real problem is that `ValidateLogin2FANewWorkspacePage` is in the history stack, maybe there is a better way for us to remove it from the stack without needing to call `dismissModal()`?

**Comment by tgolen** [comment]
I removed the hold because this actually needs to be deployed first, before the web-e change.

File: src/languages/en.js [hunk 2]
```diff
@@ -267,6 +267,7 @@ export default {
     },
     passwordForm: {
         pleaseFillOutAllFields: 'Please fill out all fields',
+        EnterATwoFactorAuthenticationCodeToContinue: 'Enter a two factor authentication code to continue',
```
Comment (alex-mechler) [inline src/languages/en.js:270]:
```suggestion
        enterATwoFactorAuthenticationCodeToContinue: 'Enter a two factor authentication code to continue',
```

File: src/pages/ValidateLogin2FANewWorkspacePage.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        continueSessionFromECom(accountID, validateCode);
+    }
+
+    validateAndSubmitForm() {
+        if (!this.state.twoFactorAuthCode.trim()) {
+            this.setState({formError: this.props.translate('passwordForm.pleaseFillOutAllFields')});
+            return;
+        }
+
+
```
Comment (alex-mechler) [inline src/pages/ValidateLogin2FANewWorkspacePage.js:79]: Extra blank line here

**Review by alex-mechler — CHANGES_REQUESTED** [review]

File: src/pages/ValidateLogin2FANewWorkspacePage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply src/pages/ValidateLogin2FANewWorkspacePage.js:59]: >maybe there is a better way for us to remove it from the stack without needing to call dismissModal()

I'm having trouble working out a better way. It feels like there are contexts where you'd want the default `navigate()` method to preserve the underlying modal in the stack e.g. moving from `/settings` to `/settings/payments` and contexts where you'll want a different behavior. We could maybe add a parameter to `navigate()` that forces all other modals to close?

[the remark this answers is not in the document]
Comment (tgolen) [reply src/pages/ValidateLogin2FANewWorkspacePage.js:59]: All right, I was able to fix this. Turns out, I was navigating to that modal twice (once in Session, and again from ComponentDidMount (from when the same page was remounted: once in `public screens` and then again in `auth screens`).

Once I removed the navigation in the Session action, it all worked like a charm.

**Comment by tgolen** [comment]
All right, thanks for the reviews! I updated it and addressed everything as well as fixing the initial warnings that I was struggling with. Should be ready for another review.

File: src/libs/API.js [hunk 4]
```diff
@@ -296,6 +295,33 @@ function reauthenticate(command = '') {
...
+/**
+ * Calls the command=Authenticate API with an accountID, validateCode, and optional 2FA code. This is used specifically
+ * for sharing sessions between e.com and this app. It will return an authToken that is used for initiating a session
+ * in this app. This API call doesn't have any special handling (like retries or special error handling).
+ *
+ * @param {Object} parameters
+ * @param {String} parameters.accountID
+ * @param {String} parameters.validateCode
+ * @param {String} [parameters.twoFactorAuthCode]
+ * @returns {Promise<unknown>}
```
Comment (marcaaron) [inline src/libs/API.js:307]: NAB / just curious but why `Promise<unknown>`?

File: src/libs/Navigation/AppNavigator/AuthScreens.js [hunk 5]
```diff
@@ -93,6 +95,14 @@ const modalScreenListeners = {
     },
 };
 
+const defaultScreenOptions = {
+    cardStyle: {
+        overflow: 'visible',
+    },
+    headerShown: false,
+    animationTypeForReplace: 'pop',
+};
```
Comment (marcaaron) [inline src/libs/Navigation/AppNavigator/AuthScreens.js:104]: Seems like `defaultScreenOptions` is [also declared in `PublicScreens.js` with the same values](https://github.com/Expensify/Expensify.cash/blob/36a336400efc060df2ebc21d74ca1ee19ee5afff/src/libs/Navigation/AppNavigator/PublicScreens.js#L10-L16). Maybe they should be moved into a separate config file so we don't have to remember to update things twice in the future when applying a change?

File: src/libs/API.js (hunk 4 above)
Comment (tgolen) [reply src/libs/API.js:307]: I think PHPStorm did that automatically, so I'm not sure why! I'll remove it

File: src/libs/Navigation/AppNavigator/AuthScreens.js (hunk 5 above)
Comment (tgolen) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:104]: Sure, I can do that.

**Comment by tgolen** [comment]
Updated to DRY up the default screen options.

**Comment by alex-mechler** [comment]
I'm trying to test this out, and Web-E redirects me to the correct link, but then I am redirect to the homepage without any error messages. Is there any additional setup I have to do that is not mentioned in the testing steps?

**Review by Dal-Papa — DISMISSED** [review]

**Comment by tgolen** [comment]
@alex-mechler I don't think there is anything special you need to do, no.

Just so I am clear, the test steps work for except that instead of seeing the new-workspace page, you see the home page?

**Comment by alex-mechler** [comment]
Correct. I am not automatically logged into e.cash, and do not see the new workspace page. I checked out the branch from https://github.com/Expensify/Web-Expensify/pull/31302 and this branch. Is there any other branches I need to checkout? Was there any auth changes that I might be missing?

File: src/languages/en.js [hunk 6]
```diff
@@ -268,6 +268,7 @@ export default {
     },
     passwordForm: {
         pleaseFillOutAllFields: 'Please fill out all fields',
+        enterATwoFactorAuthenticationCodeToContinue: 'Enter a two factor authentication code to continue',
```
Comment (alex-mechler) [inline src/languages/en.js:271]: NAB: I feel like `your` reads better than `a` here

File: src/pages/ValidateLogin2FANewWorkspacePage.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                            onSubmitEditing={this.validateAndSubmitForm}
+                            keyboardType={CONST.KEYBOARD_TYPE.NUMERIC}
+                        />
+                    </View>
+                    <View>
+                        <Button
+                            success
+                            style={[styles.mb2]}
+                            text={this.props.translate('common.continue')}
+                            isLoading={this.state.loading}
```
Comment (alex-mechler) [inline src/pages/ValidateLogin2FANewWorkspacePage.js:121]: NAB: We never set this anywhere but the constructor, meaning the user has no feedback as to when we are actually loading

**Review by alex-mechler — DISMISSED** [review]
Turns out my auth was out of date 😅 

Works well, two NABs, feel free to merge if you don't think they are worth fixing

File: src/languages/en.js (hunk 6 above)
Comment (tgolen) [reply src/languages/en.js:271]: OK, I can change this.

File: src/pages/ValidateLogin2FANewWorkspacePage.js (hunk 7 above)
Comment (tgolen) [reply src/pages/ValidateLogin2FANewWorkspacePage.js:121]: I'll see about cleaning this up.

**Comment by tgolen** [comment]
OK, I updated to change the language on the form prompt and to display the loading indicator when the 2FA is being validated.

**Review by alex-mechler — APPROVED** [review]

**Comment by tgolen** [comment]
I'm gonna go ahead and merge since the other reviewers approved already and the changes since then were minor.

**Comment by tgolen** [comment]
Once this goes to production, we can merge Expensify/Web-Expensify#31302

**Comment by kavimuru** [comment]
##  User is not automatically logged into e.cash after completing the inbox task
## Expected Result: 
Scenario 1: ​Verify that you are automatically logged into e.cash and are seeing the page to create a new workspace
Scenario 2:  Verify that you are redirected to e.cash to ever a 2FA code
## Actual Result: 
Not automatically logged into e.cash and the page to create a new workspace isn't shown. User is redirected to sign in page for both the scenarios
## Actions Performed:
Scenario1:
1. Be sure you're signed out of e.cash
2. Sign into e.com with a brand new account on a domain WITHOUT two-factor authentication
3. Complete the inbox task "What will you use Expensify for? Choose as many as you'd like!" and ONLY choose the expensify card

Scenario2:
1. Be sure you're signed out of e.cash
2. Sign into e.com with a brand new account on a domain WITH two-factor authentication enabled
3. Complete the inbox task "What will you use Expensify for? Choose as many as you'd like!" and ONLY choose the expensify card
## Platform:
Web ✔️
iOS 
mWeb 
Android 
## Build:
1.0.77-0
## Notes/Images/Video:
Bug5147298_Screen_Shot_2021-07-10_at_7 30 21_PM

Bug5147298_Screen_Shot_2021-07-10_at_7 35 37_PM

**Comment by tgolen** [comment]
Ah, I think that's expected for now. Sorry, this issue should be production QA and internal QA (because it requires being in a beta for it to work). It also needs to have https://github.com/Expensify/Web-Expensify/pull/31302 deployed, and it looks like that still hasn't been.

**Comment by isagoico** [comment]
Perfect! will check the PR off the list then. Thanks 🙇

**Comment by TomatoToaster** [comment]
ITS WORKING!!
I got redirected from expensify after gettting the new task
image

**Comment by tgolen** [comment]
Yeah! Noice

**Pull request merged**